### PR TITLE
Twitch OAuth2

### PR DESCRIPTION
**Pull request by AnatoleAM**

### CODE REVIEW

File: src/externalapis/externalapis.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+// ReadRequestResponse: quick utility for decoding an api response to a struct
+func ReadRequestResponse(resp *http.Response, out interface{}) error {
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	// cock and balls
+	fmt.Println(string(b))
```
Comment (TroyDota) [inline src/externalapis/externalapis.go:27]: Tssk

**Review by TroyDota — CHANGES_REQUESTED** [review]
Remove it WeirdChamp

File: src/configure/collections.go [hunk 2]
```diff
@@ -23,8 +22,7 @@ var Indexes = []mongo.IndexRef{
 	{
 		Collection: CollectionNameUsers,
 		Index: mongo.IndexModel{
-			Keys:    bson.M{"username": 1},
-			Options: options.Index().SetUnique(true),
+			Keys: bson.M{"username": 1},
```
Comment (TroyDota) [inline src/configure/collections.go:25]: you will have to manually change this index.

File: src/server/v3/authentication/twitch.oauth2.go [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+		if len(csrfToken) != 3 {
+			return helpers.HttpResponse(c).
+				SetMessage(fmt.Sprintf("Bad State (found %d segments when 3 were expected)", len(csrfToken))).
+				SetStatus(helpers.HttpStatusCodeBadRequest).
+				SendAsError()
+		}
+
+		// Verify the token
+		var csrfClaim *auth.JWTClaimOAuth2CSRF
+		token, _, err := auth.VerifyJWT(gCtx.Config().Credentials.JWTSecret, csrfToken)
```
Comment (TroyDota) [inline src/server/v3/authentication/twitch.oauth2.go:93]: does this work?

File: src/server/v3/authentication/twitch.oauth2.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			logrus.WithError(err).Error("twitch")
+			return helpers.HttpResponse(c).
+				SetMessage("Internal Request Rejected by External Provider").SetStatus(helpers.HttpStatusCodeInternalServerError).SendAsError()
+		}
+		defer resp.Body.Close()
+
+		var grant *OAuth2AuthorizedResponse
+		if err = externalapis.ReadRequestResponse(resp, &grant); err != nil {
```
Comment (TroyDota) [inline src/server/v3/authentication/twitch.oauth2.go:167]: This shouldnt work.

File: src/server/v3/authentication/twitch.oauth2.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+			// Add the connection to user object
+			ub.AddConnection(connection.ID)
+
+			// Find user
+			doc = gCtx.Inst().Mongo.Collection(mongo.CollectionNameUsers).FindOne(ctx, bson.M{
+				"connections": bson.M{
+					"$in": []primitive.ObjectID{connection.ID},
+				},
+			})
+			if err = doc.Decode(&user); err == mongo.ErrNoDocuments {
```
Comment (TroyDota) [inline src/server/v3/authentication/twitch.oauth2.go:215]: Neither should this.

**Review by TroyDota — CHANGES_REQUESTED** [review]

File: src/configure/collections.go (hunk 2 above)
Comment (AnatoleAM) [reply src/configure/collections.go:25]: That's fine this isn't deployed anywhere, I already did so locally

File: src/server/v3/authentication/twitch.oauth2.go (hunk 4 above)
Comment (AnatoleAM) [reply src/server/v3/authentication/twitch.oauth2.go:167]: I believe so, what would be wrong with it?

File: src/server/v3/authentication/twitch.oauth2.go (hunk 5 above)
Comment (AnatoleAM) [reply src/server/v3/authentication/twitch.oauth2.go:215]: What's wrong with that?

File: src/server/v3/authentication/twitch.oauth2.go (hunk 4 above)
Comment (TroyDota) [reply src/server/v3/authentication/twitch.oauth2.go:167]: remove the * from the `var grant *OAuth2AuthorizedResponse`

File: src/server/v3/authentication/twitch.oauth2.go (hunk 5 above)
Comment (TroyDota) [reply src/server/v3/authentication/twitch.oauth2.go:215]: The same as above, you are taking the pointer of a pointer where the pointer is null. This is wrong, you should do 

user := &User{}
and then unmarshal without passing the pointer 
ie. `.Decode(user)`

File: src/server/v3/authentication/twitch.oauth2.go (hunk 4 above)
Comment (TroyDota) [reply src/server/v3/authentication/twitch.oauth2.go:167]: or define the grant liek this

```go
grant := &OAuth2AuthorizedResponse{}
if err = externalapis.ReadRequestResponse(resp, grant); err != nil {
```

File: src/server/v3/authentication/twitch.oauth2.go (hunk 3 above)
Comment (TroyDota) [reply src/server/v3/authentication/twitch.oauth2.go:93]: Same as below.

File: src/externalapis/externalapis.go (hunk 1 above)
Comment (TroyDota) [reply src/externalapis/externalapis.go:27]: remove fmt.Println

**Pull request merged**